File: PolicyPlus/RegFile.vb
Imports System.IO
Imports Microsoft.Win32
Public Class RegFile
    ' This class implements just enough of IPolicySource to work with PolFile.ApplyDifference
    ' It is not a valid policy source for any policy loader
    Implements IPolicySource
    Const RegSignature = "Windows Registry Editor Version 5.00"
    Dim Prefix As String ' REG files require fully-rooted key paths, while other policy sources disallow them
    Dim SourceSubtree As String ' Accept only writes under this policy path (not needed if only going to use Apply)
    Dim Keys As New List(Of RegFileKey)
    Private Shared Function EscapeValue(Text As String) As String
        ' Escape quotes and slashes for a value name or string data
        Dim sb As New Text.StringBuilder
        For n = 0 To Text.Length - 1
            Dim character = Text(n)
            If character = """"c Or character = "\"c Then sb.Append("\"c)
            sb.Append(character)
        Next
        Return sb.ToString
    End Function
    Private Shared Function UnescapeValue(Text As String) As String
        ' The reverse of EscapeValue
        Dim sb As New Text.StringBuilder
        Dim escaping As Boolean = False
        For n = 0 To Text.Length - 1
            If escaping Then
                sb.Append(Text(n))
                escaping = False
            ElseIf Text(n) = "\"c Then
                escaping = True
            Else
                sb.Append(Text(n))
            End If
        Next
        Return sb.ToString
    End Function
    Private Shared Function ReadNonCommentingLine(Reader As StreamReader, Optional StopAt As Char? = Nothing) As String
        Do
            If Reader.EndOfStream Then Return Nothing
            If StopAt.HasValue AndAlso Reader.Peek = Asc(StopAt.Value) Then Return Nothing
            Dim line = Reader.ReadLine
            If String.IsNullOrWhiteSpace(line) OrElse Trim(line).StartsWith(";") Then Continue Do
            Return line
        Loop
    End Function
    Public Shared Function Load(Reader As StreamReader, Prefix As String) As RegFile
        If Reader.ReadLine() <> RegSignature Then Throw New InvalidDataException("Incorrect REG signature")
        Dim reg As New RegFile
        reg.SetPrefix(Prefix)
        Do ' Read all the keys
            Dim keyHeader = ReadNonCommentingLine(Reader)
            If keyHeader Is Nothing Then Exit Do
            Dim keyName = keyHeader.Substring(1, keyHeader.Length - 2) ' Remove the brackets
            If keyName.StartsWith("-") Then
                ' It's a deleter
                Dim deleterKey As New RegFileKey With {.Name = keyName.Substring(1), .IsDeleter = True}
                reg.Keys.Add(deleterKey)
            Else
                Dim key As New RegFileKey With {.Name = keyName}
                Do ' Read all the values
                    Dim valueLine = ReadNonCommentingLine(Reader, "["c)
                    If valueLine Is Nothing Then Exit Do
                    Dim valueName As String = ""
                    Dim data As String
                    If valueLine.StartsWith("@") Then
                        data = valueLine.Substring(2)
                    Else
                        Dim parts = Split(valueLine, """=", 2)
                        valueName = UnescapeValue(parts(0).Substring(1))
                        data = parts(1)
                    End If
                    Dim value As New RegFileValue With {.Name = valueName}
                    If data = "-" Then
                        value.IsDeleter = True
                    ElseIf data.StartsWith("""") Then
                        value.Kind = RegistryValueKind.String
                        value.Data = UnescapeValue(data.Substring(1, data.Length - 2))
                    ElseIf data.StartsWith("dword:") Then
                        value.Kind = RegistryValueKind.DWord
                        value.Data = UInteger.Parse(data.Substring(6), Globalization.NumberStyles.HexNumber)
                    ElseIf data.StartsWith("hex") Then
                        Dim indexOfClosingParen = data.IndexOf(")"c)
                        Dim curHexLine As String
                        If indexOfClosingParen <> -1 Then
                            value.Kind = Integer.Parse(data.Substring(4, indexOfClosingParen - 4), Globalization.NumberStyles.HexNumber)
                            curHexLine = data.Substring(indexOfClosingParen + 2)
                        Else
                            value.Kind = RegistryValueKind.Binary
                            curHexLine = data.Substring(4)
                        End If
                        Dim allDehexedBytes As New List(Of Byte)
                        Do ' Read all the hex lines
                            Dim hexBytes = Split(Trim(curHexLine).TrimEnd("\"c, ","c), ",").Where(Function(s) s <> "")
                            For Each b In hexBytes
                                allDehexedBytes.Add(Byte.Parse(b, Globalization.NumberStyles.HexNumber))
                            Next
                            If curHexLine.EndsWith("\") Then
                                curHexLine = Reader.ReadLine
                            Else
                                Exit Do
                            End If
                        Loop
                        value.Data = PolFile.BytesToObject(allDehexedBytes.ToArray, value.Kind)
                    End If
                    key.Values.Add(value)
                Loop
                reg.Keys.Add(key)
            End If
        Loop
        Return reg
    End Function
    Public Shared Function Load(File As String, Prefix As String) As RegFile
        Using reader As New StreamReader(File)
            Return Load(reader, Prefix)
        End Using
    End Function
    Public Sub Save(Writer As StreamWriter)
        Writer.WriteLine(RegSignature)
        Writer.WriteLine()
        For Each key In Keys
            If key.IsDeleter Then
                Writer.WriteLine("[-" & key.Name & "]")
            Else
                Writer.WriteLine("[" & key.Name & "]")
                For Each value In key.Values
                    Dim posInRow = 0 ' To split hex across lines
                    If value.Name = "" Then
                        Writer.Write("@")
                        posInRow = 1
                    Else
                        Dim quotedName = """" & EscapeValue(value.Name) & """"
                        Writer.Write(quotedName)
                        posInRow = quotedName.Length
                    End If
                    Writer.Write("=")
                    posInRow += 1
                    If value.IsDeleter Then
                        Writer.WriteLine("-")
                    Else
                        Select Case value.Kind
                            Case RegistryValueKind.String
                                Writer.Write("""")
                                Writer.Write(EscapeValue(value.Data))
                                Writer.WriteLine("""")
                            Case RegistryValueKind.DWord
                                Writer.Write("dword:")
                                Writer.WriteLine(Convert.ToString(CUInt(value.Data), 16).PadLeft(8, "0"c))
                            Case Else
                                Writer.Write("hex")
                                posInRow += 3
                                If value.Kind <> RegistryValueKind.Binary Then
                                    Writer.Write("(")
                                    Writer.Write(Convert.ToString(value.Kind, 16))
                                    Writer.Write(")")
                                    posInRow += 3
                                End If
                                Writer.Write(":")
                                posInRow += 1
                                Dim bytes = PolFile.ObjectToBytes(value.Data, value.Kind)
                                For n = 0 To bytes.Length - 2
                                    Writer.Write(Convert.ToString(bytes(n), 16).PadLeft(2, "0"c))
                                    Writer.Write(",")
                                    posInRow += 3
                                    If posInRow >= 78 Then
                                        Writer.WriteLine("\")
                                        Writer.Write("  ")
                                        posInRow = 2
                                    End If
                                Next
                                If bytes.Length > 0 Then
                                    Writer.WriteLine(Convert.ToString(bytes(bytes.Length - 1), 16).PadLeft(2, "0"c))
                                Else
                                    Writer.WriteLine()
                                End If
                        End Select
                    End If
                Next
            End If
            Writer.WriteLine()
        Next
    End Sub
    Public Sub Save(File As String)
        Using writer As New StreamWriter(File, False)
            Save(writer)
        End Using
    End Sub
    Private Function UnprefixKeyName(Name As String) As String
        If Name.StartsWith(Prefix, StringComparison.InvariantCultureIgnoreCase) Then Return Name.Substring(Prefix.Length) Else Return Name
    End Function
    Private Function PrefixKeyName(Name As String) As String
        Return Prefix & Name
    End Function
    Private Function GetKey(Name As String) As RegFileKey
        Return Keys.FirstOrDefault(Function(k) k.Name.Equals(Name, StringComparison.InvariantCultureIgnoreCase))
    End Function
    Private Function GetKeyByUnprefixedName(Name As String) As RegFileKey
        Return GetKey(PrefixKeyName(Name))
    End Function
    Private Function GetOrCreateKey(Name As String) As RegFileKey
        Dim key = GetKey(Name)
        If key Is Nothing Then
            key = New RegFileKey With {.Name = Name}
            Keys.Add(key)
        End If
        Return key
    End Function
    Private Function GetNonDeleterKey(Name As String) As RegFileKey
        Return Keys.FirstOrDefault(Function(k) (Not k.IsDeleter) AndAlso k.Name.Equals(Name, StringComparison.InvariantCultureIgnoreCase))
    End Function
    Private Function IsSourceKeyAcceptable(Key As String) As Boolean
        Return (SourceSubtree = "") OrElse Key.Equals(SourceSubtree, StringComparison.InvariantCultureIgnoreCase) OrElse
            Key.StartsWith(SourceSubtree & "\", StringComparison.InvariantCultureIgnoreCase)
    End Function
    Public Function ContainsValue(Key As String, Value As String) As Boolean Implements IPolicySource.ContainsValue
        Throw New NotImplementedException()
    End Function
    Public Function GetValue(Key As String, Value As String) As Object Implements IPolicySource.GetValue
        Throw New NotImplementedException()
    End Function
    Public Function WillDeleteValue(Key As String, Value As String) As Boolean Implements IPolicySource.WillDeleteValue
        Throw New NotImplementedException()
    End Function
    Public Function GetValueNames(Key As String) As List(Of String) Implements IPolicySource.GetValueNames
        Throw New NotImplementedException()
    End Function
    Public Sub SetValue(Key As String, Value As String, Data As Object, DataType As RegistryValueKind) Implements IPolicySource.SetValue
        If Not IsSourceKeyAcceptable(Key) Then Exit Sub
        Dim fullKeyName = PrefixKeyName(Key)
        Dim keyRecord = GetNonDeleterKey(fullKeyName)
        If keyRecord Is Nothing Then
            keyRecord = New RegFileKey With {.Name = fullKeyName}
            Keys.Add(keyRecord)
        End If
        keyRecord.Values.Remove(keyRecord.GetValue(Value))
        keyRecord.Values.Add(New RegFileValue With {.Name = Value, .Kind = DataType, .Data = Data})
    End Sub
    Public Sub ForgetValue(Key As String, Value As String) Implements IPolicySource.ForgetValue
        Throw New NotImplementedException()
    End Sub
    Public Sub DeleteValue(Key As String, Value As String) Implements IPolicySource.DeleteValue
        If Not IsSourceKeyAcceptable(Key) Then Exit Sub
        Dim fullKeyName = PrefixKeyName(Key)
        Dim keyRecord = GetOrCreateKey(fullKeyName)
        If keyRecord.IsDeleter Then Exit Sub
        keyRecord.Values.Remove(keyRecord.GetValue(Value))
        keyRecord.Values.Add(New RegFileValue With {.Name = Value, .IsDeleter = True})
    End Sub
    Public Sub ClearKey(Key As String) Implements IPolicySource.ClearKey
        If Not IsSourceKeyAcceptable(Key) Then Exit Sub
        Dim fullName = PrefixKeyName(Key)
        Keys.Remove(GetKey(fullName))
        Keys.Add(New RegFileKey With {.Name = fullName, .IsDeleter = True})
    End Sub
    Public Sub ForgetKeyClearance(Key As String) Implements IPolicySource.ForgetKeyClearance
        If Not IsSourceKeyAcceptable(Key) Then Exit Sub
        Dim keyRecord = GetKeyByUnprefixedName(Key)
        If keyRecord Is Nothing Then Exit Sub
        If keyRecord.IsDeleter Then Keys.Remove(keyRecord)
    End Sub
    Public Sub Apply(Target As IPolicySource)
        For Each key In Keys
            Dim unprefixedKeyName = UnprefixKeyName(key.Name)
            If key.IsDeleter Then
                Target.ClearKey(unprefixedKeyName)
            Else
                For Each value In key.Values
                    If value.IsDeleter Then
                        Target.DeleteValue(unprefixedKeyName, value.Name)
                    Else
                        Target.SetValue(unprefixedKeyName, value.Name, value.Data, value.Kind)
                    End If
                Next
            End If
        Next
    End Sub
    Public Sub SetPrefix(Prefix As String)
        If Not Prefix.EndsWith("\") Then Prefix &= "\"
        Me.Prefix = Prefix
    End Sub
    Public Sub SetSourceBranch(Branch As String)
        If Branch.EndsWith("\") Then Branch = Branch.TrimEnd("\"c)
        SourceSubtree = Branch
    End Sub
    Public Function GuessPrefix() As String
        ' Try to determine a reasonable prefix from the data present
        If Keys.Count = 0 Then Return "HKEY_LOCAL_MACHINE\" ' Can't do much without any data
        Dim firstKeyName = Keys(0).Name
        If firstKeyName.StartsWith("HKEY_USERS\") Then
            ' The user SID should be part of the prefix
            Dim secondSlashPos = firstKeyName.IndexOf("\", 11)
            Return firstKeyName.Substring(0, secondSlashPos + 1)
        Else
            ' Other hives should be just fine
            Dim firstSlashPos = firstKeyName.IndexOf("\")
            Return firstKeyName.Substring(0, firstSlashPos + 1)
        End If
    End Function
    Public Function HasDefaultValues() As Boolean
        Return Keys.Any(Function(k) k.Values.Any(Function(v) v.Name = ""))
    End Function
    Private Class RegFileKey
        Public Name As String
        Public IsDeleter As Boolean
        Public Values As New List(Of RegFileValue)
        Public Function GetValue(Value As String) As RegFileValue
            Return Values.FirstOrDefault(Function(v) v.Name.Equals(Value, StringComparison.InvariantCultureIgnoreCase))
        End Function
    End Class
    Private Class RegFileValue
        Public Name As String
        Public Data As Object
        Public Kind As RegistryValueKind
        Public IsDeleter As Boolean
    End Class
End Class

File: PolicyPlus/RegFile.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;
using Microsoft.Win32;

namespace PolicyPlus
{
    // This class implements just enough of IPolicySource to work with PolFile.ApplyDifference
    // It is not a valid policy source for any policy loader
    public class RegFile : IPolicySource
    {
        private const string RegSignature = "Windows Registry Editor Version 5.00";
        private string Prefix; // REG files require fully-rooted key paths, while other policy sources disallow them
        private string SourceSubtree; // Accept only writes under this policy path (not needed if only going to use Apply)
        private List<RegFileKey> Keys = new List<RegFileKey>();
        private static string EscapeValue(string Text)
        {
            // Escape quotes and slashes for a value name or string data
            var sb = new System.Text.StringBuilder();
            for (int n = 0, loopTo = Text.Length - 1; n <= loopTo; n++)
            {
                char character = Text[n];
                if (character == '"' | character == '\\')
                    sb.Append('\\');
                sb.Append(character);
            }
            return sb.ToString();
        }
        private static string UnescapeValue(string Text)
        {
            // The reverse of EscapeValue
            var sb = new System.Text.StringBuilder();
            bool escaping = false;
            for (int n = 0, loopTo = Text.Length - 1; n <= loopTo; n++)
            {
                if (escaping)
                {
                    sb.Append(Text[n]);
                    escaping = false;
                }
                else if (Text[n] == '\\')
                {
                    escaping = true;
                }
                else
                {
                    sb.Append(Text[n]);
                }
            }
            return sb.ToString();
        }
        private static string ReadNonCommentingLine(StreamReader Reader, char? StopAt = default)
        {
            do
            {
                if (Reader.EndOfStream)
                    return null;
                if (StopAt.HasValue && Reader.Peek() == Strings.Asc(StopAt.Value))
                    return null;
                string line = Reader.ReadLine();
                if (string.IsNullOrWhiteSpace(line) || Strings.Trim(line).StartsWith(";"))
                    continue;
                return line;
            }
            while (true);
        }
        public static RegFile Load(StreamReader Reader, string Prefix)
        {
            if ((Reader.ReadLine() ?? "") != RegSignature)
                throw new InvalidDataException("Incorrect REG signature");
            var reg = new RegFile();
            reg.SetPrefix(Prefix);
            do // Read all the keys
            {
                string keyHeader = ReadNonCommentingLine(Reader);
                if (keyHeader is null)
                    break;
                string keyName = keyHeader.Substring(1, keyHeader.Length - 2); // Remove the brackets
                if (keyName.StartsWith("-"))
                {
                    // It's a deleter
                    var deleterKey = new RegFileKey() { Name = keyName.Substring(1), IsDeleter = true };
                    reg.Keys.Add(deleterKey);
                }
                else
                {
                    var key = new RegFileKey() { Name = keyName };
                    do // Read all the values
                    {
                        string valueLine = ReadNonCommentingLine(Reader, '[');
                        if (valueLine is null)
                            break;
                        string valueName = "";
                        string data;
                        if (valueLine.StartsWith("@"))
                        {
                            data = valueLine.Substring(2);
                        }
                        else
                        {
                            string[] parts = Strings.Split(valueLine, "\"=", 2);
                            valueName = UnescapeValue(parts[0].Substring(1));
                            data = parts[1];
                        }
                        var value = new RegFileValue() { Name = valueName };
                        if (data == "-")
                        {
                            value.IsDeleter = true;
                        }
                        else if (data.StartsWith("\""))
                        {
                            value.Kind = RegistryValueKind.String;
                            value.Data = UnescapeValue(data.Substring(1, data.Length - 2));
                        }
                        else if (data.StartsWith("dword:"))
                        {
                            value.Kind = RegistryValueKind.DWord;
                            value.Data = uint.Parse(data.Substring(6), System.Globalization.NumberStyles.HexNumber);
                        }
                        else if (data.StartsWith("hex"))
                        {
                            int indexOfClosingParen = data.IndexOf(')');
                            string curHexLine;
                            if (indexOfClosingParen != -1)
                            {
                                value.Kind = (RegistryValueKind)int.Parse(data.Substring(4, indexOfClosingParen - 4), System.Globalization.NumberStyles.HexNumber);
                                curHexLine = data.Substring(indexOfClosingParen + 2);
                            }
                            else
                            {
                                value.Kind = RegistryValueKind.Binary;
                                curHexLine = data.Substring(4);
                            }
                            var allDehexedBytes = new List<byte>();
                            do // Read all the hex lines
                            {
                                var hexBytes = Strings.Split(Strings.Trim(curHexLine).TrimEnd('\\', ','), ",").Where(s => !string.IsNullOrEmpty(s));
                                foreach (var b in hexBytes)
                                    allDehexedBytes.Add(byte.Parse(b, System.Globalization.NumberStyles.HexNumber));
                                if (curHexLine.EndsWith(@"\"))
                                {
                                    curHexLine = Reader.ReadLine();
                                }
                                else
                                {
                                    break;
                                }
                            }
                            while (true);
                            value.Data = PolFile.BytesToObject(allDehexedBytes.ToArray(), value.Kind);
                        }
                        key.Values.Add(value);
                    }
                    while (true);
                    reg.Keys.Add(key);
                }
            }
            while (true);
            return reg;
        }
        public static RegFile Load(string File, string Prefix)
        {
            using (var reader = new StreamReader(File))
            {
                return Load(reader, Prefix);
            }
        }
        public void Save(StreamWriter Writer)
        {
            Writer.WriteLine(RegSignature);
            Writer.WriteLine();
            foreach (var key in Keys)
            {
                if (key.IsDeleter)
                {
                    Writer.WriteLine("[-" + key.Name + "]");
                }
                else
                {
                    Writer.WriteLine("[" + key.Name + "]");
                    foreach (var value in key.Values)
                    {
                        int posInRow = 0; // To split hex across lines
                        if (string.IsNullOrEmpty(value.Name))
                        {
                            Writer.Write("@");
                            posInRow = 1;
                        }
                        else
                        {
                            string quotedName = "\"" + EscapeValue(value.Name) + "\"";
                            Writer.Write(quotedName);
                            posInRow = quotedName.Length;
                        }
                        Writer.Write("=");
                        posInRow += 1;
                        if (value.IsDeleter)
                        {
                            Writer.WriteLine("-");
                        }
                        else
                        {
                            switch (value.Kind)
                            {
                                case RegistryValueKind.String:
                                    {
                                        Writer.Write("\"");
                                        Writer.Write(EscapeValue(Conversions.ToString(value.Data)));
                                        Writer.WriteLine("\"");
                                        break;
                                    }
                                case RegistryValueKind.DWord:
                                    {
                                        Writer.Write("dword:");
                                        Writer.WriteLine(Convert.ToString(Conversions.ToUInteger(value.Data), 16).PadLeft(8, '0'));
                                        break;
                                    }

                                default:
                                    {
                                        Writer.Write("hex");
                                        posInRow += 3;
                                        if (value.Kind != RegistryValueKind.Binary)
                                        {
                                            Writer.Write("(");
                                            Writer.Write(Convert.ToString((int)value.Kind, 16));
                                            Writer.Write(")");
                                            posInRow += 3;
                                        }
                                        Writer.Write(":");
                                        posInRow += 1;
                                        byte[] bytes = PolFile.ObjectToBytes(value.Data, value.Kind);
                                        for (int n = 0, loopTo = bytes.Length - 2; n <= loopTo; n++)
                                        {
                                            Writer.Write(Convert.ToString(bytes[n], 16).PadLeft(2, '0'));
                                            Writer.Write(",");
                                            posInRow += 3;
                                            if (posInRow >= 78)
                                            {
                                                Writer.WriteLine(@"\");
                                                Writer.Write("  ");
                                                posInRow = 2;
                                            }
                                        }
                                        if (bytes.Length > 0)
                                        {
                                            Writer.WriteLine(Convert.ToString(bytes[bytes.Length - 1], 16).PadLeft(2, '0'));
                                        }
                                        else
                                        {
                                            Writer.WriteLine();
                                        }

                                        break;
                                    }
                            }
                        }
                    }
                }
                Writer.WriteLine();
            }
        }
        public void Save(string File)
        {
            using (var writer = new StreamWriter(File, false))
            {
                Save(writer);
            }
        }
        private string UnprefixKeyName(string Name)
        {
            if (Name.StartsWith(Prefix, StringComparison.InvariantCultureIgnoreCase))
                return Name.Substring(Prefix.Length);
            else
                return Name;
        }
        private string PrefixKeyName(string Name)
        {
            return Prefix + Name;
        }
        private RegFileKey GetKey(string Name)
        {
            return Keys.FirstOrDefault(k => k.Name.Equals(Name, StringComparison.InvariantCultureIgnoreCase));
        }
        private RegFileKey GetKeyByUnprefixedName(string Name)
        {
            return GetKey(PrefixKeyName(Name));
        }
        private RegFileKey GetOrCreateKey(string Name)
        {
            var key = GetKey(Name);
            if (key is null)
            {
                key = new RegFileKey() { Name = Name };
                Keys.Add(key);
            }
            return key;
        }
        private RegFileKey GetNonDeleterKey(string Name)
        {
            return Keys.FirstOrDefault(k => !k.IsDeleter && k.Name.Equals(Name, StringComparison.InvariantCultureIgnoreCase));
        }
        private bool IsSourceKeyAcceptable(string Key)
        {
            return string.IsNullOrEmpty(SourceSubtree) || Key.Equals(SourceSubtree, StringComparison.InvariantCultureIgnoreCase) || Key.StartsWith(SourceSubtree + @"\", StringComparison.InvariantCultureIgnoreCase);
        }
        public bool ContainsValue(string Key, string Value)
        {
            throw new NotImplementedException();
        }
        public object GetValue(string Key, string Value)
        {
            throw new NotImplementedException();
        }
        public bool WillDeleteValue(string Key, string Value)
        {
            throw new NotImplementedException();
        }
        public List<string> GetValueNames(string Key)
        {
            throw new NotImplementedException();
        }
        public void SetValue(string Key, string Value, object Data, RegistryValueKind DataType)
        {
            if (!IsSourceKeyAcceptable(Key))
                return;
            string fullKeyName = PrefixKeyName(Key);
            var keyRecord = GetNonDeleterKey(fullKeyName);
            if (keyRecord is null)
            {
                keyRecord = new RegFileKey() { Name = fullKeyName };
                Keys.Add(keyRecord);
            }
            keyRecord.Values.Remove(keyRecord.GetValue(Value));
            keyRecord.Values.Add(new RegFileValue() { Name = Value, Kind = DataType, Data = Data });
        }
        public void ForgetValue(string Key, string Value)
        {
            throw new NotImplementedException();
        }
        public void DeleteValue(string Key, string Value)
        {
            if (!IsSourceKeyAcceptable(Key))
                return;
            string fullKeyName = PrefixKeyName(Key);
            var keyRecord = GetOrCreateKey(fullKeyName);
            if (keyRecord.IsDeleter)
                return;
            keyRecord.Values.Remove(keyRecord.GetValue(Value));
            keyRecord.Values.Add(new RegFileValue() { Name = Value, IsDeleter = true });
        }
        public void ClearKey(string Key)
        {
            if (!IsSourceKeyAcceptable(Key))
                return;
            string fullName = PrefixKeyName(Key);
            Keys.Remove(GetKey(fullName));
            Keys.Add(new RegFileKey() { Name = fullName, IsDeleter = true });
        }
        public void ForgetKeyClearance(string Key)
        {
            if (!IsSourceKeyAcceptable(Key))
                return;
            var keyRecord = GetKeyByUnprefixedName(Key);
            if (keyRecord is null)
                return;
            if (keyRecord.IsDeleter)
                Keys.Remove(keyRecord);
        }
        public void Apply(IPolicySource Target)
        {
            foreach (var key in Keys)
            {
                string unprefixedKeyName = UnprefixKeyName(key.Name);
                if (key.IsDeleter)
                {
                    Target.ClearKey(unprefixedKeyName);
                }
                else
                {
                    foreach (var value in key.Values)
                    {
                        if (value.IsDeleter)
                        {
                            Target.DeleteValue(unprefixedKeyName, value.Name);
                        }
                        else
                        {
                            Target.SetValue(unprefixedKeyName, value.Name, value.Data, value.Kind);
                        }
                    }
                }
            }
        }
        public void SetPrefix(string Prefix)
        {
            if (!Prefix.EndsWith(@"\"))
                Prefix += @"\";
            this.Prefix = Prefix;
        }
        public void SetSourceBranch(string Branch)
        {
            if (Branch.EndsWith(@"\"))
                Branch = Branch.TrimEnd('\\');
            SourceSubtree = Branch;
        }
        public string GuessPrefix()
        {
            // Try to determine a reasonable prefix from the data present
            if (Keys.Count == 0)
                return @"HKEY_LOCAL_MACHINE\"; // Can't do much without any data
            string firstKeyName = Keys[0].Name;
            if (firstKeyName.StartsWith(@"HKEY_USERS\"))
            {
                // The user SID should be part of the prefix
                int secondSlashPos = firstKeyName.IndexOf(@"\", 11);
                return firstKeyName.Substring(0, secondSlashPos + 1);
            }
            else
            {
                // Other hives should be just fine
                int firstSlashPos = firstKeyName.IndexOf(@"\");
                return firstKeyName.Substring(0, firstSlashPos + 1);
            }
        }
        public bool HasDefaultValues()
        {
            return Keys.Any(k => k.Values.Any(v => string.IsNullOrEmpty(v.Name)));
        }
        private class RegFileKey
        {
            public string Name;
            public bool IsDeleter;
            public List<RegFileValue> Values = new List<RegFileValue>();
            public RegFileValue GetValue(string Value)
            {
                return Values.FirstOrDefault(v => v.Name.Equals(Value, StringComparison.InvariantCultureIgnoreCase));
            }
        }
        private class RegFileValue
        {
            public string Name;
            public object Data;
            public RegistryValueKind Kind;
            public bool IsDeleter;
        }
    }
}